connection_producer.go

```diff
@@ -0,0 +1,218 @@
+)
+
+type snowflakeConnectionProducer struct {
+	ConnectionURL         string `json:"connection_url"`
+	MaxOpenConnections    int    `json:"max_open_connections"`
+	MaxIdleConnections    int    `json:"max_idle_connections"`
+	MaxConnectionLifetime string `json:"max_connection_lifetime"`
+	Username              string `json:"username"`
+	Password              string `json:"password"`
+	PrivateKey            string `json:"private_key"`
+	UsernameTemplate      string `json:"username_template"`
+	DisableEscaping       bool   `json:"disable_escaping"`
+
+	Initialized bool
+	RawConfig   map[string]any
+	Type        string
+	snowflakeDB *sql.DB
+	sync.Mutex
+}
+
+func (c *snowflakeConnectionProducer) secretValues() map[string]string {
+	return map[string]string{
+		c.Password: "[password]",
+	}
+}
+
+func (c *snowflakeConnectionProducer) Init(ctx context.Context, initConfig map[string]interface{}, verifyConnection bool) (saveConfig map[string]interface{}, err error) {
+	c.Lock()
+	defer c.Unlock()
+
+	c.RawConfig = initConfig
+
+	decoderConfig := &mapstructure.DecoderConfig{
+		Result:           c,
+		WeaklyTypedInput: true,
+		TagName:          "json",
+	}
+
+	decoder, err := mapstructure.NewDecoder(decoderConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	err = decoder.Decode(initConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(c.Password) > 0 {
+		// Return an error here once Snowflake ends support for password auth.
+	}
+
+	c.Initialized = true
+
+	if verifyConnection {
+		if _, err := c.Connection(ctx); err != nil {
+			c.close()
+			return nil, fmt.Errorf("error verifying connection: %w", err)
+		}
+	}
+
+	return initConfig, nil
+}
+
+func (c *snowflakeConnectionProducer) Initialize(ctx context.Context, config map[string]any, verifyConnection bool) error {
+	_, err := c.Init(ctx, config, verifyConnection)
+	return err
+}
+
+func (c *snowflakeConnectionProducer) Connection(ctx context.Context) (interface{}, error) {
+	// This is intentionally not grabbing the lock since the calling functions (e.g. CreateUser)
+	// are claiming it. (The locking patterns could be refactored to be more consistent/clear.)
+
+	if !c.Initialized {
+		return nil, connutil.ErrNotInitialized
+	}
+
+	if c.snowflakeDB != nil {
+		return c.snowflakeDB, nil
+	}
+
+	var db *sql.DB
+	var err error
+	if c.PrivateKey != "" {
+		db, err = openSnowflake(c.ConnectionURL, c.Username, c.PrivateKey)
+		if err != nil {
+			return nil, fmt.Errorf("error opening Snowflake connection: %w", err)
+		}
+	} else {
+		db, err = sql.Open(snowflakeSQLTypeName, c.ConnectionURL)
+		if err != nil {
+			return nil, fmt.Errorf("failed to open connection: %w", err)
+		}
+	}
+
+	c.snowflakeDB = db
+
+	return db, nil
+}
+
+// close terminates the database connection without locking
+func (c *snowflakeConnectionProducer) close() error {
+	if c.snowflakeDB != nil {
+		if err := c.snowflakeDB.Close(); err != nil {
+			return err
+		}
+	}
+
+	c.snowflakeDB = nil
```

Comment:
Should we move this to the locked `Close()` call below? We may run into a race condition otherwise.